File: vlc-android/src/org/videolan/vlc/gui/network/MRLAdapter.java
/*****************************************************************************
 * MRLAdapter.java
 *****************************************************************************
 * Copyright © 2014-2015 VLC authors and VideoLAN
 * Author: Geoffrey Métais
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/
package org.videolan.vlc.gui.network;

import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.videolan.medialibrary.media.HistoryItem;
import org.videolan.medialibrary.media.MediaWrapper;
import org.videolan.vlc.R;

class MRLAdapter extends RecyclerView.Adapter<MRLAdapter.ViewHolder> {
    private HistoryItem[] mDataset;
    private MediaPlayerController playerController;

    interface MediaPlayerController {
        void playMedia(MediaWrapper mw);
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        final TextView uriTv, titleTv;

        public ViewHolder(View v) {
            super(v);
            uriTv = v.findViewById(R.id.mrl_item_uri);
            titleTv = v.findViewById(R.id.mrl_item_title);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            playerController.playMedia(mDataset[getLayoutPosition()].getMedia());
        }
    }

    MRLAdapter(MediaPlayerController playerController) {
        this.playerController = playerController;
    }

    @Override
    public MRLAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
                                                    int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.mrl_item, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        final HistoryItem item = mDataset[position];
        holder.uriTv.setText(Uri.decode(item.getMrl()));
        holder.titleTv.setText(Uri.decode(item.getTitle()));
    }

    public void setList(HistoryItem[] list){
        mDataset = list;
        notifyDataSetChanged();
    }

    public HistoryItem getItem(int position) {
        if (position >= getItemCount() || position < 0)
            return null;
        return mDataset[position];
    }

    @Override
    public int getItemCount() {
        return mDataset.length;
    }

    public boolean isEmpty(){
        return getItemCount() == 0;
    }
}

File: vlc-android/src/org/videolan/vlc/gui/network/MRLAdapter.kt
/*****************************************************************************
 * MRLAdapter.java
 *
 * Copyright © 2014-2015 VLC authors and VideoLAN
 * Author: Geoffrey Métais
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
package org.videolan.vlc.gui.network

import androidx.databinding.DataBindingUtil
import android.net.Uri
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.coroutines.channels.SendChannel
import org.videolan.medialibrary.media.MediaWrapper
import org.videolan.vlc.R
import org.videolan.vlc.databinding.MrlItemBinding

internal class MRLAdapter(private val eventActor: SendChannel<MrlAction>) : androidx.recyclerview.widget.RecyclerView.Adapter<MRLAdapter.ViewHolder>() {
    private var dataset: Array<MediaWrapper>? = null

    val isEmpty: Boolean
        get() = itemCount == 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MRLAdapter.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding: MrlItemBinding = DataBindingUtil.inflate(inflater, R.layout.mrl_item, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = dataset?.get(position)
        holder.binding.mrlItemUri.text = Uri.decode(item?.location)
        holder.binding.mrlItemTitle.text = Uri.decode(item?.title)
    }

    fun setList(list: Array<MediaWrapper>?) {
        dataset = list
        notifyDataSetChanged()
    }

    fun getItem(position: Int): MediaWrapper? = when {
        position >= itemCount -> null
        position < 0 -> null
        else -> dataset?.get(position)
    }

    override fun getItemCount() = dataset?.size ?: 0

    inner class ViewHolder(val binding: MrlItemBinding) : androidx.recyclerview.widget.RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        init {
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener { eventActor.offer(ShowContext(layoutPosition)) }
            binding.mrlCtx.setOnClickListener { eventActor.offer(ShowContext(layoutPosition)) }
        }

        override fun onClick(v: View) {
            dataset?.get(layoutPosition)?.let { eventActor.offer(Playmedia(it)) }
        }
    }
}

sealed class MrlAction
class Playmedia(val media: MediaWrapper) : MrlAction()
class ShowContext(val position: Int) : MrlAction()
